### DISCUSSION TITLE
Conflict with a ground net

### DISCUSSION THREAD

**Discussion by gabrielrodcanal** [Xilinx/fpga24_routing_contest#54] [Q&A] [answered]
Hi,
I have generated a simple benchmark to test my router, which is based on the custom Python router provided in the contest. I'm working on fixing overlaps, since this is not included in the base version, and I've found that I get collisions with GND signals. How is this possible? Aren't GND and VCC nets removed from the routing graph in the beginning of the program? In particular, Vivado complains there's a collision between one of my generated nets and GLOBAL_LOGIC0. When I inspect the conflicts with `show_objects -name find_1 [get_nets -hierarchical -top_net_of_hierarchical_group -filter { ROUTE_STATUS == "CONFLICTS" } ]` I get the following:
![image](https://github.com/Xilinx/fpga24_routing_contest/assets/23747903/692dfd57-19f3-48f6-8367-0d2b574ebd58)

Any guidance here would be appreciated.

**Comment by eddieh-xlnx** [comment]
Hi @gabrielrodcanal. Thanks for posting!

> I've found that I get collisions with GND signals. How is this possible? Aren't GND and VCC nets removed from the routing graph in the beginning of the program?

For `nxroute-poc`, no effort is made to remove or block all the routing resources used by the pre-routed GND/VCC/clock nets. There is, however, this block of code:

https://github.com/Xilinx/fpga24_routing_contest/blob/e8e0ee81aad38e38ce12a3c91eeb41a2dd93c1da/networkx-proof-of-concept-router/nxroute-poc.py#L384-L391

that is responsible for blocking the routing resources of just the *site pin* nodes (but not the actual routing nodes that connects site pins together) on those pre-routed nets. The reason for this is that Vivado has (at least) two types of conflicts: site pin conflicts and route conflicts. Site pin conflicts means that a net is overlapping with a site pin from another net -- and it's obvious who should "win" that resource, since the site pin is the only way for the latter net to reach the fabric. For a route conflict, it's less obvious (from Vivado's point of view) who really deserves that resource the most. However, from the point of view of this contest, static/pre-routed nets **must** be preserved and thus your router will need to avoid those nodes.

To keep things simple for this MVP nxroute, I prevent site pin conflicts from ever happening (code above) and so what you'll only ever see are route conflicts.

Long story short, you will need to remove the routing resources used by pre-routed nets from the graph yourself after which those routing conflicts should go away.

**Reply by gabrielrodcanal** [reply]
Hi @eddieh-xlnx. Thank you for your answer.

I guess it would be enough with removing just the sources and the sinks of the pre-routed nets instead of all their nodes in the graph?

**Reply by eddieh-xlnx** [reply]
> I guess it would be enough with removing just the sources and the sinks of the pre-routed nets instead of all their nodes in the graph?

No, what I'm saying is that I currently remove the source/sink nodes of pre-routed nets from the graph. As you've found, that's not enough because to-be-routed nets can cause "route conflicts" with pre-routed nets. You'll have to go ahead and remove all pre-routed net nodes from the graph too.

**Reply by eddieh-xlnx** [reply, marked as answer]
> You'll have to go ahead and remove all pre-routed net nodes from the graph too.

Resolved in https://github.com/Xilinx/fpga24_routing_contest/pull/55
